sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Error related to the Simulator interface. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "EMU error: {}".  So including the words "error" or
+/// "Emu" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct EmuArgs {
+    param: String,
+    value: Option<String>,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+pub enum EmuState {
+    Off,
+    On,
+    Busy,
+    Error,
+}
+
+/// A trait which represents a single GPIO pin.
+pub trait Emulator {
+    /// Aquire status of emulator
+    fn status(&self) -> Result<EmuState>;
```

Comment:
In the interest of reducing protocol chatter, I wonder whether it would make sense to decide that this method delays returning until the emulator is out of `Busy` state.  That is, `Busy` would then not even be a possible value in the enum, only an internal condition in (some) emulators.

I am bringing this up because I foresee callers wanting to do something like below, which would be inefficient, particularly if the transport goes via the network.
```
emulator.start(...)?;
loop {
  sleep(Duration::from_millis(100); 
  match emulator.status()? {
    EmuState::Busy => continue;
    EmuState::On => break,
    _ => bail!("Emulator failed to start"),
  }
}
```

If the host emulator can remain busy for tens of seconds, then maybe my proposal to drop the `Busy` state does not make sense.  Still, I think the blocking version would be useful, so in that case, maybe two version of `status()` could be warranted.

Reply:
Busy state exist because other traits like GPIO or I2C should not perform operation during start/stop cycle.
Also Emulator itself shouldn't allow of running two start in the same time.

I think it is possible to remove busy state but we need to add time out to functions.
The problem is with error... Error need to be extended with "device_busy" error or something like that.

Reply:
I see.  I suppose we need a new Error::DeviceBusy no matter how we choose to represent the state enum, for cases when e.g. Gpio operation is requested while the emulator is starting up.

Then maybe we can do with a single function like this:
```
fn get_state(&self, timeout: Duration) -> Result<EmuState>;
```
in which busy would be indicated by `Err(Busy)`, rather than an explicit enum value in `State`.  Callers that want an immediate answer can give zero timeout, and use `match` to see if the emulator is busy.  Others can give a long timeout, and would probably use the operator `?`, as they do not need to gracefully handle the case of the emulator remaining busy through the entire timeout.

Reply:
New code contain function wait_for_state(...)
If we have 2 client (A) and (B) we can make something like this:
(A)
emulator start timeout=3000 ....

(B)
// wait_for_state(5000, EmuState::On)
emulator wait_for_state timeout=5000 state=On
gpio write LED_RED 1